### PR TITLE
Add handlers to get xep-0359 (Unique and Stable Stanza IDs)

### PR DESCRIPTION
**Pull request by mazhe**
Hi,

I've been using a bit shmoose in conjonction with a home server and a spectrum gateway. It seems that spectrum especially send messages without an "id" attribute when MaM is enabled, maybe because they consider the xep-0359 (Unique and Stable Stanza IDs) element is enough (revised xep-313 does indeed mandate a stanza-id element that is the canonical message id).

Anyway, a message without messageid will be ignored by shmoose because it will not have a key to be stored in persitent store.

Swift do not handle well stanza-id elements, but I've followed https://swift.im/swiften/guide/#Section-Extending to create a very simple handling/payload declaration of stanza-id elements, and prefer it as a "messageid" if present.

What do you think of this? I guess some refactoring might be needed if you want, I've tried to follow the styling I saw in the rest of the project.

Thanks for doing a promising native xmpp client for  sailfish!

### CODE REVIEW

**Comment by geobra** [comment]
Hi

Thank you for the PR and working on the code to make shmoose better.

Two thoughts on this topic:

I also encountered that some clients did not include an id field in the message part. You are right, in this case shmoose just drops that message which is a bad behaviour.

I had a quick look at xep-0359. It's state is deferred, so I don't know if this is the best solution of the problem. And even if we decide to go this way, the overall code has to be adopted on more places. The message ID is also used for message-marker (sent, received, read state). And probably on other functions as well.

Please give me some days to go more into this topic and let me think about how to proceed here...

**Comment by mazhe** [comment]
Thanks!

Yeah, I saw that it state was deferred, but I guessed that since it is specified as a mandatory part of MaM (which itself is also still experimental), it was more of a way to improve MaM support in the long run and fix this more immediate "messages without ids" issue.

Please do take your time to consider this indeed, there might be some cases where I've missed the "receive" part, and indeed that mean there should be something to do about the "send" part.

I also missed that support can be discovered from entities, instead of just trying to check if there's an element...

Have a nice day!

**Comment by geobra** [comment]
I did some reading in the relevant xeps.

I am a big fan of implementing stuff iterative by use-case instead of implementing complete xeps at once. As far as I can tell, your use-case is to make Shmoose work to also accept messages without an ID field and instead use xep-0359 for this. In my opinion, the implementation should be started exactly with this. For now, I can't figure out any advantages for MaM or room messaging, but maybe I didn't read that part in the other xep's in that detail.

So please:
- Use your implementation for that use-case. I will write some notes into your code
- Don't use disco info to announce the support of this xep, as it is only partly implemented
- Don't ask the other party for this xep for now, just try to get an ID with xep 359 if no ID is in the msg field
- Update the github start page with XEP-0359 [WIP]
- Use an underscore at the end of private members (like privatMember_)
- Use an own line for the curly brackets of methods
- Initialize your private members in the C++11 way (like auto privateMember_{};). This will be reworked in the complete source

What is your opinion on this?

File: source/base/MessageHandler.cpp [hunk 1]
```diff
@@ -109,15 +110,27 @@ void MessageHandler::handleMessageReceived(Swift::Message::ref message)
             isGroupMessage = true;
         }
 
+        // xep 0359
```
Comment (geobra) [inline source/base/MessageHandler.cpp:113]: Make the message ID the default one. Only if no message ID can be found, use the StanzaID.

File: source/xep/stanzaId/StanzaId.h [hunk 2]
```diff
@@ -0,0 +1,88 @@
+#pragma once
```
Comment (geobra) [inline source/xep/stanzaId/StanzaId.h:1]: Make one file for each class. Separate the header from implementation.

File: source/base/MessageHandler.cpp (hunk 1 above)
Comment (mazhe) [reply source/base/MessageHandler.cpp]: Easy, will do!

File: source/xep/stanzaId/StanzaId.h (hunk 2 above)
Comment (mazhe) [reply source/xep/stanzaId/StanzaId.h:1]: Done, I wasn't that fond of having one header, but Swiften structure makes me declare a few relatively trivial classes and I was not sure about having 8 files in this subdirectory for a relatively minor feature...

**Comment by mazhe** [comment]
I'm fine with the requested amendments, those are minor things I wasn't sure at first (like naming/styling conventions that were coming for some from Swiften doc).

I've refactored the PR to account (I think) for everyone save the C++11 private member instanciation, not because I don't agree, but because I'm not sure to see how they are done (not really a C++ guru). I'll do another commit once I figure this one.

File: source/base/MessageHandler.cpp [hunk 3]
```diff
@@ -109,15 +110,27 @@ void MessageHandler::handleMessageReceived(Swift::Message::ref message)
             isGroupMessage = true;
         }
 
+        QString messageId = QString::fromStdString(message->getID());
+		if (messageId.length() == 0)
```
Comment (geobra) [inline source/base/MessageHandler.cpp:114]: Fix the indentation

File: source/base/MessageHandler.cpp [hunk 4]
```diff
@@ -109,15 +110,27 @@ void MessageHandler::handleMessageReceived(Swift::Message::ref message)
...
+        {
+            // No message id, try xep 0359
+            std::shared_ptr<StanzaIdPayload> stanzaId = message->getPayload<StanzaIdPayload>();
+            if (stanzaId != nullptr)
+            {
+                QString messageId = QString::fromStdString(stanzaId->getId());
+
```
Comment (geobra) [inline source/base/MessageHandler.cpp:121]: This messageId gets out of scope after the closing curly bracket.
Remove the empty line.

Comment (mazhe) [reply source/base/MessageHandler.cpp:121]: Yes, I just realized that a few minutes ago...

**Comment by geobra** [comment]
After fixing the minor findings, please make a functional manual test. 
Or (prefered!) add an integration test for this case.

**Comment by mazhe** [comment]
Hmm, I quickly looked at existing tests... unit testing would require tests for the base message handler I guess? integration testing would require the ability to inject arbitrary stanzas (since shmoose will always set a id attribute to a message)...

What I see feasible with my limited knowledge of the project testing environment would be testing the swiften payload parsing part...

**Comment by geobra** [comment]
Yes, you are right. Integration testing is difficult with the current setup. It's just shmoose talking to shmoose. Not sending message id's would require huge refactoring, which is not the purpose of this PR.

To have some unit-tests of the parsing part would be great. Have a look at "test/unit_tests/MamManager" to get a knowledge about the few existing ones. Please set them up in a similar manner.

**Comment by mazhe** [comment]
Hi, ok here we go with a first throw. It is quite rudimentary, I spent a few hours trying to look at both how swift and shmoose organize their testsuites and try to put something meaningful.

What I settled with was : 
- Validate the parsing at the parser level like swift would do (since at base, it's extending swiften payloads), done with a custom parser class extending swiften's
- Validate serialization... it's unused, but swiften guide made it look it was mandatory to implement alongside a parser, so it'll be there if/when needed
- Validate the client setup... Not much to test except that the code do not segfault...

What do you think of this proposal?

**Comment by geobra** [comment]
For me it looks very good now. I have no other idea what could be additionally be tested. I will merge this PR now. Thank you for the great work.

**Pull request merged**